samza-api/src/main/java/org/apache/samza/system/SystemStreamPartition.java

```diff
@@ -38,6 +39,7 @@ public SystemStreamPartition(String system, String stream, Partition partition)
     super(system, stream);
     this.partition = partition;
     this.hash = computeHashCode();
+    this.keyBucket = -1;
```

Comment:
why not default the keyBucket to just 0 or 1 however you are basing the bucket index? Cut shorts all the additional handling of -1 and so-on and also logically makes sense that in the event of keyBucket < 1 there is one bucket to which all the of the messages belonging to the task is assigned to.

Reply:
+1, keyBucket can just be 0.

Reply:
agreed that keeping keyBucket = 0 when there is no elasticity makes this particular SSP class simpler.

my reason for having it to -1 was as follows.
making keyBucket = 0 and removing the check `keyBucket == -1` means every time SSP is printed into logs and task model, it will be `SystemStreamPartition [system, stream, partition, 0] with no elasticity enabled which can be confusing.

this check of `keyBucket-1` will be restricted to this class and hence i think it is better to have this additional check rather than having to introduce SSP containing 0 in all places. 

pl lmk if this makes sense

Reply:
during the second pr https://github.com/apache/samza/pull/1580 dev, i realized that there is a bigger problem with having keyBucket=0 in the ssp.tostring when elasticity is not enabled.

GroupBySystemStreamPartition uses ssp.tostring as the taksName and task level metrics (like sends etc) use the task name. keeping keyBucket=0 in all ssp.tostring even when elasticity is not enabled will result in loss of metrics history.

Reply:
But cant you still have the !=0 check in toString to preserve existing behavior?

Reply:
yes, that is doable but imo doesnt differ much from the current !=-1 check. pl lmk if you feel strongly about this and i can update.

Reply:
How exactly is this doable? is your bucket based on 1-index is it? when you have elasticity enabled, one of the bucket will have 0 as the index and hence its representation would be different from others.

Reply:
yes you are right.currently my indexes are 0-based and hence using -1 as the default when elasticity is off.

if i choose keyBucket=0 for elasticity off, then i will have to start my elasticity enabled indexes from 1,2.. but this does not differ from default when elasticity off = -1 and start indexes from 0,1,2..